06_gpu_and_ml/audio-to-text/parakeet.py

```diff
@@ -0,0 +1,282 @@
+# Using a secondary image for the client allows us to keep the client dependencies separate from the server dependencies.
+# We'll use  python's `websockets` library to create a websocket client
+# that sends audio data to the server and receives transcriptions in real-time.
+
+
+client_image = modal.Image.debian_slim(python_version="3.12").pip_install(
+    "websockets==15.0.1", "wave==0.0.2", "numpy==2.2.5"
+)
+client_app = modal.App("parakeet-client", image=client_image)
+
+
+WS_ENDPOINT = "/ws"
+
+AUDIO_URL = "https://github.com/voxserv/audio_quality_testing_samples/raw/refs/heads/master/mono_44100/156550__acclivity__a-dream-within-a-dream.wav"
+
+
+@client_app.local_entrypoint()
+def main(modal_profile: str, audio_url: str = AUDIO_URL):
+    import asyncio
+
+    import requests
+
+    CHUNK_SIZE = 64000
+    is_dev = True
+    url = f"wss://{modal_profile}--{app_name}-{class_name}-web{'-dev' if is_dev else ''}.modal.run"
+    ws_url = f"{url}{WS_ENDPOINT}"
+
+    def convert_to_mono_16khz(audio_bytes: bytes) -> bytes:
+        import io
+        import wave
+
+        import numpy as np
+
+        with wave.open(io.BytesIO(audio_bytes), "rb") as wav_in:
+            n_channels = wav_in.getnchannels()
+            sample_width = wav_in.getsampwidth()
+            frame_rate = wav_in.getframerate()
+            n_frames = wav_in.getnframes()
+            frames = wav_in.readframes(n_frames)
+
+        # Determine dtype from sample width
+        if sample_width == 1:
+            dtype = np.uint8
+        elif sample_width == 2:
+            dtype = np.int16
+        elif sample_width == 4:
+            dtype = np.int32
+        else:
+            raise ValueError(f"Unsupported sample width: {sample_width}")
+
+        # Convert frames to NumPy array
+        audio_data = np.frombuffer(frames, dtype=dtype)
+
+        # Downmix to mono if needed
+        if n_channels > 1:
+            audio_data = audio_data.reshape(-1, n_channels)
+            audio_data = audio_data.mean(axis=1).astype(dtype)
+
+        # Resample to 16kHz if needed
+        if frame_rate != 16000:
+            ratio = 16000 / frame_rate
+            new_length = int(len(audio_data) * ratio)
+            indices = np.linspace(0, len(audio_data) - 1, new_length)
+            audio_data = np.interp(
+                indices, np.arange(len(audio_data)), audio_data
+            ).astype(dtype)
+
+        return audio_data.tobytes()
+
+    def chunk_audio(data: bytes, chunk_size: int):
+        for i in range(0, len(data), chunk_size):
+            yield data[i : i + chunk_size]
+
+    async def send_audio(ws, audio_bytes):
+        for chunk in chunk_audio(audio_bytes, CHUNK_SIZE):
+            await ws.send(chunk)
+            await asyncio.sleep(0.01)  # simulate real-time pacing
+
+    async def receive_transcriptions(ws):
+        async for message in ws:
+            print("📝 Transcription:", message)
+
+    async def run(ws_url, audio_bytes):
+        import websockets
```

Comment:
Unfortunately I don't think that there's a stdlib implementation of `websockets`, so I'm not sure how we're going to implement this.